File: TencentCloudLogProducer/tencentcloud-log-c-sdk/src/log_producer_client.c
//
// Created by herrylv on 06/5/2022
//
#include "log_producer_client.h"
#include "log_producer_manager.h"
#include "cls_log.h"
#include "utils.h"
#include "post_logs_api.h"
#include <stdarg.h>
#include <string.h>
#include <stdlib.h>

static uint32_t s_init_flag = 0;
static int s_last_result = 0;
static int search_init_api_flag = 0;

unsigned int LOG_GET_TIME();

typedef struct
{

    ProducerManager *producermgr;
    ProducerConfig *producerconf;

} PrivateProducerClient;

struct clslogproducer
{
    clslogproducerclient *ancestor;
};

int  ClsLogProducerInit()
{
    // if already init, just return s_last_result
    if (s_init_flag == 1)
    {
        return s_last_result;
    }
    s_init_flag = 1;
    if (0 != cls_log_init(LOG_GLOBAL_ALL))
    {
        s_last_result = LOG_PRODUCER_INVALID;
    }
    else
    {
        s_last_result = LOG_PRODUCER_OK;
    }
    return s_last_result;
}

void ClsLogProducerDestroy()
{
    if (s_init_flag == 0)
    {
        return;
    }
    s_init_flag = 0;
    cls_log_destroy();
}

clslogproducer *ConstructorClsLogProducer(ProducerConfig *config, SendCallBackFunc callbackfunc, void *user_param)
{
    if (!is_valid(config))
    {
        return NULL;
    }
    clslogproducer *producer = (clslogproducer *)malloc(sizeof(clslogproducer));
    clslogproducerclient *producerclient = (clslogproducerclient *)malloc(sizeof(clslogproducerclient));
    PrivateProducerClient *privateclient = (PrivateProducerClient *)malloc(sizeof(PrivateProducerClient));
    producerclient->private_client = privateclient;
    privateclient->producerconf = config;
    privateclient->producermgr = ConstructorProducerManager(config);
    privateclient->producermgr->callbackfunc = callbackfunc;
    privateclient->producermgr->user_param = user_param;

    cls_debug_log("create producer client success, topic : %s", config->topic);
    producerclient->efficient = true;
    producer->ancestor = producerclient;
    return producer;
}

void DestructorClsLogProducer(clslogproducer *producer)
{
    if (producer == NULL)
    {
        return;
    }
    clslogproducerclient *client = producer->ancestor;
    client->efficient = false;
    PrivateProducerClient *client_private = (PrivateProducerClient *)client->private_client;
    destroy_log_producer_manager(client_private->producermgr);
    DestroyClsLogProducerConfig(client_private->producerconf);
    free(client_private);
    free(client);
    free(producer);
}

extern clslogproducerclient *GetClsLogProducer(clslogproducer *producer, const char *config_name)
{
    if (producer == NULL)
    {
        return NULL;
    }
    return producer->ancestor;
}

int
PostClsLog(clslogproducerclient *client,
           int64_t logtime,
                                                int32_t pair_count, char **keys,
                                                int32_t *key_lens, char **values,
                                                int32_t *value_lens, int flush)
{
    if (client == NULL || !client->efficient)
    {
        return LOG_PRODUCER_INVALID;
    }

    ProducerManager *manager = ((PrivateProducerClient *)client->private_client)->producermgr;
    return log_producer_manager_add_log(manager, logtime,pair_count, keys, key_lens, values, value_lens, flush, -1);
}

//search log
int ClsSearchLog(const char *region,const char *secretid, const char* secretkey,const char* logsetid,const char **topicids,const int topicidslens,const char* starttime,const char* endtime,const char* query,size_t limit,const char* context,const char* sort,get_result* result){
    //参数校验
    int iRet = SearchLogCheckParam(region,secretid, secretkey,logsetid,topicids, topicidslens, starttime, endtime,query, limit,sort);
    if(iRet != 0){
        result->statusCode = iRet;
        return iRet;
    }
    
    //构造Query数据
    root_t params = RB_ROOT;
    put(&params, "logset_id", logsetid);
    char *strtopics = malloc(4096);
    memset(strtopics,0,4096);
    arr_to_string(topicids, topicidslens, strtopics);
    put(&params, "topic_ids", strtopics);
    put(&params, "start_time", starttime);
    put(&params, "end_time", endtime);
    put(&params, "query_string", query);
    char slimit[10];
    memset(slimit,0,sizeof(slimit));
    sprintf(slimit, " %d" , limit);
    put(&params, "limit", slimit);
    if(context != NULL){
        put(&params, "context", context);
    }
    if(sort != NULL){
        put(&params, "sort", sort);
    }
    
    //构造header数据
    root_t httpHeader = RB_ROOT;
    put(&httpHeader, "Host", (char *)region);
    char c_signature[1024];
    signature(secretid, secretkey, "GET", "/searchlog", params, httpHeader, 300, c_signature);
    put(&httpHeader, "Authorization", c_signature);
    
    //发起请求
    SearchLogApi(region, httpHeader, params,result);
    free(strtopics);
    return 0;
}

int SearchLogCheckParam(const char *region,const char *secretid, const char* secretkey,const char* logsetid,const char **topicids,const int topicidslens,const char* starttime,const char* endtime,const char* query,size_t limit,const char* sort){
    if(strlen(region) == 0 || strlen(secretid) == 0 ||strlen(secretkey) == 0||strlen(logsetid) == 0||topicids == NULL ||topicidslens <= 0||strlen(starttime) == 0||strlen(endtime) == 0 ||limit == 0){
        return -1;
    }
    char *tmpsort = (char*)malloc(strlen(sort)+1);
    memset(tmpsort,0,strlen(sort)+1);
    strlowr(sort,tmpsort);
    if(sort != NULL && (strncmp(tmpsort,"asc",3) != 0)&&(strncmp(tmpsort,"desc",4) != 0)){
        return -1;
    }
    free(tmpsort);
    return 0;
}

int ClsLogSearchLogInit()
{
    if (search_init_api_flag == 1)
    {
        return s_last_result;
    }
    search_init_api_flag = 1;
    if (0 != cls_log_init(LOG_GLOBAL_ALL))
    {
        s_last_result = LOG_PRODUCER_INVALID;
    }
    else
    {
        s_last_result = LOG_PRODUCER_OK;
    }
    return s_last_result;
}

void ClsLogSearchLogDestroy()
{
    if (search_init_api_flag == 0)
    {
        return;
    }
    search_init_api_flag = 0;
    cls_log_destroy();
}

File: TencentCloudLogProducer/TencentCloudLogProducer/LogProducerClient.m


#import <Foundation/Foundation.h>
#import "LogProducerClient.h"
#import "LogProducerConfig.h"
#import "TencentCloudLogProducer/Log.h"
#import "TimeUtils.h"
#import "sds.h"



@interface LogProducerClient ()
@end

@implementation LogProducerClient

- (id) initWithClsLogProducer:(LogProducerConfig *)logProducerConfig
{
    return [self initWithClsLogProducer:logProducerConfig callback:nil];
}

- (id) initWithClsLogProducer:(LogProducerConfig *)logProducerConfig callback:(SendCallBackFunc)callback
{
    if (self = [super init])
    {
         ClsLogProducerInit(LOG_GLOBAL_ALL);
        self->producer = ConstructorClsLogProducer(logProducerConfig->config, *callback, nil);
        self->client = GetClsLogProducer(self->producer, nil);
        if(self->client == NULL){
            enable = false;
        }else{
            enable = YES;
        }
        self->logConfig = logProducerConfig;
        
    }
    return self;
}

- (void) UpdateSecurityToken:(NSString *)securityToken{
    if ([securityToken length] == 0) {
        return;
    }
    [logConfig ResetSecurityToken:securityToken];
}

- (void)DestroyLogProducer
{
    if (!enable) {
        return;
    }
    enable = NO;
    DestructorClsLogProducer(self->producer);
    ClsLogProducerDestroy();
}

- (LogProducerResult)PostLog:(Log *) log
{
    return [self PostLog:log flush:0];
}

- (LogProducerResult)PostLog:(Log *) log flush:(int) flush
{
    if (!enable || self->client == NULL || log == nil) {
        return LogProducerInvalid;
    }
    NSMutableDictionary *logContents = log->content;
    
    int pairCount = (int)[logContents count];
        
    char **keyArray = (char **)malloc(sizeof(char *)*(pairCount));
    char **valueArray = (char **)malloc(sizeof(char *)*(pairCount));
    
    int32_t *keyCountArray = (int32_t*)malloc(sizeof(int32_t)*(pairCount));
    int32_t *valueCountArray = (int32_t*)malloc(sizeof(int32_t)*(pairCount));
    
    
    int ids = 0;
    for (NSString *key in logContents) {
        NSString *value = logContents[key];

        char* keyChar=[self convertToChar:key];
        char* valueChar=[self convertToChar:value];

        keyArray[ids] = keyChar;  //记录contents 所有的key
        valueArray[ids] = valueChar; ////记录contents 所有的value
        keyCountArray[ids] = (int32_t)strlen(keyChar); //记录contents 所有的key的大小
        valueCountArray[ids] = (int32_t)strlen(valueChar); //记录contents 所有的value的大小
        
        ids = ids + 1;
    }
    
    
    int res = PostClsLog(self->client, log->logTime, pairCount, keyArray, keyCountArray, valueArray, valueCountArray, flush);
    
    for(int i=0;i<pairCount;i++) {
        free(keyArray[i]);
        free(valueArray[i]);
    }
    free(keyArray);
    free(valueArray);
    free(keyCountArray);
    free(valueCountArray);
    return res;
}

-(char*)convertToChar:(NSString*)strtemp
{
    NSUInteger len = [strtemp lengthOfBytesUsingEncoding:NSUTF8StringEncoding] + 1;
    if (len > 1000000) return strdup([strtemp UTF8String]);
    char cStr [len];
    [strtemp getCString:cStr maxLength:len encoding:NSUTF8StringEncoding];
    return strdup(cStr);
}

@end



@implementation LogSearchClient
-(SearchReult) SearchLog:(NSString*)region secretid:(NSString*) secretid
             secretkey:(NSString*) secretkey
              logsetid:(NSString*) logsetid
              topicids:(NSArray*) topicids
             starttime:(NSString*) starttime
               endtime:(NSString*) endtime
                 query:(NSString*) query
                 limit:(NSInteger)limit
               context:(NSString*)context
                  sort:(NSString*)sort
{
    SearchReult result;
    char **topics = (char**)malloc(sizeof(char*)*128);
    for(int i = 0; i < topicids.count; i++){
        topics[i] = [topicids[i] UTF8String];
    }
    get_result r;
    memset(r.requestID, 0, 128);
    ClsSearchLog([region UTF8String],[secretid UTF8String] ,[secretkey UTF8String],[logsetid UTF8String],topics,topicids.count,[starttime UTF8String],[endtime UTF8String],[query UTF8String],limit,[context UTF8String],[sort UTF8String],&r);
    
    free(topics);
    result.statusCode = r.statusCode;
    result.message = r.message ? [NSString stringWithUTF8String:r.message] : nil;
    result.requestID = r.requestID ? [NSString stringWithUTF8String:r.requestID] : nil;
    free(r.message);
    return result;
}

-(id)init{
    self = [super init];
    ClsLogSearchLogInit();
    return self;
}
-(void) DestroyLogSearch{
    ClsLogSearchLogDestroy();
}
@end
